consul/server_manager/server_manager.go

```diff
@@ -215,36 +266,126 @@ func (sm *ServerManager) NotifyFailedServer(server *server_details.ServerDetails
-// consul servers and provides a guarantee that the order of the server list
-// isn't related to the age at which the node was added to the cluster.
-// Elsewhere we rely on the position in the server list as a hint regarding
-// the stability of a server relative to its position in the server list.
-// Servers at or near the front of the list are more stable than servers near
-// the end of the list.  Unhealthy servers are removed when serf notices the
-// server has been deregistered.
+// RebalanceServers shuffles the list of servers on this agent.  The server
+// at the front of the list is selected for the next RPC.  RPC calls that
+// fail for a particular server are rotated to the end of the list.  This
+// method reshuffles the list periodically in order to redistribute work
+// across all known consul servers (i.e. guarantee that the order of servers
+// in the server list isn't positively correlated with the age of a server in
+// the consul cluster).  Periodically shuffling the server list prevents
+// long-lived clients from fixating on long-lived servers.
+//
+// Unhealthy servers are removed when serf notices the server has been
+// deregistered.  Before the newly shuffled server list is saved, the new
+// remote endpoint is tested to ensure its responsive.
 func (sm *ServerManager) RebalanceServers() {
-	sm.serverConfigLock.Lock()
-	defer sm.serverConfigLock.Unlock()
-	serverCfg := sm.getServerConfig()
+FAILED_SERVER_DURING_REBALANCE:
+	// Obtain a copy of the server config
+	sc := sm.getServerConfig()
+
+	// Early abort if there is no value to shuffling
+	if len(sc.servers) < 2 {
+		// sm.logger.Printf("[INFO] server manager: can't rebalance with only %d servers", len(sc.servers))
+		return
+	}
 
-	newServers := make([]*server_details.ServerDetails, len(serverCfg.servers))
-	copy(newServers, serverCfg.servers)
+	sc.shuffleServers()
+
+	// Iterate through the shuffled server list to find a healthy server.
+	// Don't iterate on the list directly, this loop mutates server the
+	// list.
+	var foundHealthyServer bool
+	for n := len(sc.servers); n > 0; n-- {
+		// Always test the first server.  Failed servers are cycled
+		// while Serf detects the node has failed.
+		selectedServer := sc.servers[0]
+
+		// sm.logger.Printf("[INFO] server manager: Preemptively testing server %s before rebalance", selectedServer.String())
+		ok := sm.connPoolPinger.PingConsulServer(selectedServer)
+		if ok {
+			foundHealthyServer = true
+			break
+		}
+		sc.cycleServer()
+	}
 
-	// Shuffle the server list
-	for i := len(serverCfg.servers) - 1; i > 0; i-- {
-		j := rand.Int31n(int32(i + 1))
-		newServers[i], newServers[j] = newServers[j], newServers[i]
+	// If no healthy servers were found, sleep and wait for Serf to make
+	// the world a happy place again.
+	if !foundHealthyServer {
+		const backoffDuration = 1 * time.Second
+		// sm.logger.Printf("[INFO] server manager: No servers available, sleeping for %v", backoffDuration)
```

Comment:
Client's can't trigger this log line, it can only happen on a timer.  I had this here as a stub for someone and to remind about possibly using an exponential backoff logger.